docs/software_requirements/timers.sdoc

```diff
@@ -26,12 +23,334 @@ UID: ZEP-SRS-4-2
+- TYPE: Parent
+  VALUE: ZEP-SYRS-18
+
+[REQUIREMENT]
+UID: ZEP-SRS-4-18
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Timers
+TITLE: Timer thread synchronization
+STATEMENT: >>>
+The Zephyr RTOS shall provide a mechanism to synchronize a thread with a defined and initialized timer.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-SYRS-18
+
+[REQUIREMENT]
+UID: ZEP-SRS-4-19
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Timers
+TITLE: Timer expire when thread synchronization is used
+STATEMENT: >>>
+When the thread synchronization mechanism is used, the thread shall be blocked until the timer expires.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-SYRS-18
+
+[REQUIREMENT]
+UID: ZEP-SRS-4-20
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Timers
+TITLE: Timer is stopped when thread synchronization is used
+STATEMENT: >>>
+When the thread synchronization mechanism is used, the thread shall be blocked until the timer is stopped.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-SYRS-18
+
+[REQUIREMENT]
+UID: ZEP-SRS-4-21
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Timers
+TITLE: Timer thread synchronization status reset
+STATEMENT: >>>
+When the thread synchronization mechanism is used, the status of the timer shall be reset.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-SYRS-18
+
+[REQUIREMENT]
+UID: ZEP-SRS-4-22
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Timers
+TITLE: Timer remaining expiration system tick time
+STATEMENT: >>>
+The Zephyr RTOS shall provide a mechanism to get the timer remaining until expiration time in system ticks.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-SYRS-18
+
+[REQUIREMENT]
+UID: ZEP-SRS-4-23
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Timers
+TITLE: Timer remaining expiration time
+STATEMENT: >>>
+The Zephyr RTOS shall provide a mechanism to get the timer remaining until expiration time in milliseconds.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-SYRS-18
+
+[REQUIREMENT]
+UID: ZEP-SRS-4-24
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Timers
+TITLE: Timer remaining expiration time
+STATEMENT: >>>
+The Zephyr RTOS shall provide a mechanism to get the timer remaining until expiration time in milliseconds.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-SYRS-18
+
+[REQUIREMENT]
+UID: ZEP-SRS-4-25
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Timers
+TITLE: Timer set user data
+STATEMENT: >>>
+The Zephyr RTOS shall provide a mechanism to set user data to a timer.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-SYRS-18
+
+[REQUIREMENT]
+UID: ZEP-SRS-4-26
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Timers
+TITLE: Timer get user data
+STATEMENT: >>>
+The Zephyr RTOS shall provide a mechanism to get user data from a timer.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-SYRS-18
+
+[REQUIREMENT]
+UID: ZEP-SRS-4-27
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Timers
+TITLE: Timer expire functions in interrupt context
+STATEMENT: >>>
+When the timer expiry function is defined, it shall be called in the interrupt context.
```

Comment:
**question:** Is this really a requirement? To me this looks like a (natural) design rationale. This might not even be true for all architectures, ie. those that don't have a systick interrupt context like native_sim?

Reply:
In my opinion this is a requirement because if this wouldn't be here wouldn't the timer behave different on different architectures?
And do we really need to consider native sim, I don't think so because it is a simulation of a system and that usual behaves often different than a real architecture.

Reply:
Perhaps native_sim is a bad example, still systick is something cortex-m specific and I am honestly unsure if something similar exists for x86 which as a target would be in scope, not?

That said, I merely wanted to point things out I noticed, yet not with the intention to block things ...

Reply:
> Perhaps native_sim is a bad example, still systick is something cortex-m specific and I am honestly unsure if something similar exists for x86 which as a target would be in scope, not?
> 
> That said, I merely wanted to point things out I noticed, yet not with the intention to block things ...

I sneak peaked a little bit in the design and implementation and there is stated that the expiry function is executed by the system clock interrupt handler